### PR TITLE
return batches as dicts

### PR DESCRIPTION
**Pull request by themattinthehatt**
Instead of returning batches as lists, return them as dicts so that access is cleaner
images = batch["images"]
instead of 
images = batch[0]

Data loaders now also return the image index

### CODE REVIEW

**Review by danbider — CHANGES_REQUESTED** [review]
minor typechecking suggestions. looks great with dicts.

File: pose_est_nets/datasets/datasets.py [hunk 1]
```diff
@@ -112,7 +112,7 @@ def width(self):
     def __len__(self) -> int:
         return len(self.image_names)
 
-    def __getitem__(self, idx: int) -> Tuple[torch.tensor, torch.tensor]:
+    def __getitem__(self, idx: int) -> dict:
```
[the remark this answers is not in the document]
Comment (danbider) [reply pose_est_nets/datasets/datasets.py:115]: Done

File: pose_est_nets/datasets/datasets.py [hunk 2]
```diff
@@ -243,17 +251,17 @@ def compute_heatmaps(self):
         self.label_heatmaps = torch.from_numpy(np.asarray(label_heatmaps)).float()
         self.label_heatmaps = self.label_heatmaps.permute(0, 3, 1, 2)
 
-    def __getitem__(self, idx: int) -> Tuple[torch.tensor, torch.tensor, torch.Tensor]:
+    def __getitem__(self, idx: int) -> dict:
```
[the remark this answers is not in the document]
Comment (danbider) [reply pose_est_nets/datasets/datasets.py:254]: Done

File: pose_est_nets/datasets/datasets.py [hunk 3]
```diff
@@ -227,13 +235,13 @@ def compute_heatmaps(self):
         """
         label_heatmaps = []
         for idx in range(len(self.image_names)):
-            x, y = super().__getitem__(idx)
+            batch = super().__getitem__(idx)
```
[the remark this answers is not in the document]
Comment (danbider) [reply pose_est_nets/datasets/datasets.py:238]: Done

File: pose_est_nets/datasets/preprocessing.py [hunk 4]
```diff
@@ -56,20 +56,20 @@ def compute_multiview_pca_params(
         if data_module.dataset.imgaug_transform:
             i = 0
             for idx in indxs:
-                vals = pca_data.__getitem__(idx)
-                data_arr[i] = vals[1].reshape(-1, 2)
+                batch = pca_data.__getitem__(idx)
```
[the remark this answers is not in the document]
Comment (danbider) [reply pose_est_nets/datasets/preprocessing.py:59]: Done, switched to example dict

File: pose_est_nets/models/heatmap_tracker.py [hunk 5]
```diff
@@ -206,49 +206,46 @@ def forward(
         return heatmaps
 
     @typechecked
-    def training_step(self, data_batch: List, batch_idx: int) -> dict:
-        images, true_heatmaps, true_keypoints = data_batch  # read batch
-        predicted_heatmaps = self.forward(images)  # images -> heatmaps
+    def training_step(self, data_batch: Dict, batch_idx: int) -> Dict:
```
[the remark this answers is not in the document]
Comment (danbider) [reply pose_est_nets/models/heatmap_tracker.py]: done, better

File: pose_est_nets/models/regression_tracker.py [hunk 6]
```diff
@@ -173,33 +175,43 @@ def __init__(
...
         print(semi_super_losses_to_use)
+        if semi_super_losses_to_use is None:
+            raise ValueError("must specify losses for unlabeled frames")
+        if "unimodal" in semi_super_losses_to_use:
+            raise ValueError("cannot use unimodal loss in regression tracker")
         self.loss_function_dict = get_losses_dict(semi_super_losses_to_use)
         self.loss_params = loss_params
 
     @typechecked
     def training_step(self, data_batch: dict, batch_idx: int) -> dict:
```
[the remark this answers is not in the document]
Comment (danbider) [reply pose_est_nets/models/regression_tracker.py:195]: done

File: tests/test_subpixmaxima.py [hunk 7]
```diff
@@ -108,12 +119,17 @@ def test_generate_keypoints_weird_shape():
         header_rows=[1, 2],
         imgaug_transform=imgaug_transform,
     )
-    test_img, gt_heatmap, gt_keypts = dataset.__getitem__(idx=0)
-    assert(gt_keypts.shape == (torch.Size([34])))
-    gt_heatmap = gt_heatmap.unsqueeze(0)
-    gt_keypts = gt_keypts.unsqueeze(0).reshape(1, 17, 2)
+    batch = dataset.__getitem__(idx=0)
```
[the remark this answers is not in the document]
Comment (danbider) [reply tests/test_subpixmaxima.py:122]: done

File: tests/test_subpixmaxima.py [hunk 8]
```diff
@@ -30,33 +30,30 @@ def test_subpixmaxima():
         threshold=None,
         device=_TORCH_DEVICE
     )
-    test_img, gt_heatmap, gt_keypts = dataset.__getitem__(idx=0)
-    maxima, confidence = SubPixMax.run(gt_heatmap.unsqueeze(0).to(_TORCH_DEVICE))
+    batch = dataset.__getitem__(idx=0)
```
[the remark this answers is not in the document]
Comment (danbider) [reply tests/test_subpixmaxima.py:33]: done

File: pose_est_nets/utils/plotting_utils.py [hunk 9]
```diff
@@ -407,8 +407,25 @@ def predict_videos(
...
+                header_rows = [0, 1, 2]
+            df = pd.read_csv(csv_file, header=header_rows)
+            # collect marker names from multiindex header
+            joint_labels = [c[0] for c in df.columns[1::2]]
+        else:
+            joint_labels = ["bp_%i" % n for n in range(model.num_keypoints)]
+
+        # build data frame
         xyl_labels = ["x", "y", "likelihood"]
-        joint_labels = ["bp_%i" % n for n in range(model.num_keypoints)]
```
[the remark this answers is not in the document]
Comment (danbider) [reply pose_est_nets/utils/plotting_utils.py:411]: leaving it for now

[the remark this answers is not in the document]
Comment (themattinthehatt) [reply pose_est_nets/utils/plotting_utils.py:411]: no, dlc uses the name of the network, but we should leave to be consistent - maybe we can fill this out with more informative info later

**Review by themattinthehatt — COMMENTED** [review]
Very solid PR

**Pull request merged**